qiskit/passmanager/base_pass.py

```diff
@@ -0,0 +1,80 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2023.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""Metaclass of Qiskit pass manager pass."""
+
+from abc import abstractmethod
+from collections.abc import Hashable
+from inspect import signature
+from typing import Any
+
+from .propertyset import PropertySet
+
+
+class MetaPass(type):
+    """Metaclass for transpiler passes.
+
+    Enforces the creation of some fields in the pass while allowing passes to
+    override ``__init__``.
+    """
+
+    def __call__(cls, *args, **kwargs):
+        pass_instance = type.__call__(cls, *args, **kwargs)
+        pass_instance._hash = hash(MetaPass._freeze_init_parameters(cls, args, kwargs))
+        return pass_instance
+
+    @staticmethod
+    def _freeze_init_parameters(class_, args, kwargs):
+        self_guard = object()
+        init_signature = signature(class_.__init__)
+        bound_signature = init_signature.bind(self_guard, *args, **kwargs)
+        arguments = [("class_.__name__", class_.__name__)]
+        for name, value in bound_signature.arguments.items():
+            if value == self_guard:
+                continue
+            if isinstance(value, Hashable):
+                arguments.append((name, type(value), value))
+            else:
+                arguments.append((name, type(value), repr(value)))
+        return frozenset(arguments)
+
+
+class BasePass(metaclass=MetaPass):
+    """Base class for transpiler passes."""
+
+    def __init__(self):
+        self.requires = []  # List of passes that requires
+        self.preserves = []  # List of passes that preserves
```

Comment:
My best recollection is that this attribute has not been useful in practice. In fact, as far as I can see, it is not used at all in terra. In contrast, `requires` is used a bit. If it were not for backward compatibility, I imagine it would be removed. So I question whether it is a good idea to include this as a feature in the more abstract pass manager. Is it possible to define this attribute instead in `BasePass(CommonBasePass)` below?

Reply:
I think for this PR it's ok to preserve `preserves` as this is strictly for migrating the common code that already exists into a reusable module. There is a second PR #10127 built on top of this which aims to clean things up. I think we can discuss specifics like that there. I think it would be easier to view this PR as just a code migration for the ease of review. We're not locked into the api here until the final release anyway so there is time to change it post merge

Reply:
Thanks @jlapeyre , I agree with you. I also observed that options dictionary is carried around passes, 
https://github.com/Qiskit/qiskit-terra/blob/924702e9ddb3f6f0124bc8acb285ac3d1da12aff/qiskit/transpiler/runningpassmanager.py#L153
but eventually no one use this information (at least in Terra passes). However, I cannot decide what should be removed. I also want to cleanup and drop un-used features. Let's discuss carefully in #10127 . Once this is merged I'll rebase the PR for easier review.

Reply:
Sure it's ok, as long as it doesn't fall through the cracks. Btw. I think one or two people wrote or told me that `preserves` was an idea copied from llvm without knowing whether it would be useful.